File: apps/api/app/models/order.py
from sqlalchemy import CheckConstraint, Column, String, Integer, Float, Numeric, DateTime, Boolean, JSON, Enum, ForeignKey, text
from sqlalchemy.sql import func
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import relationship
from app.db.base import Base
import enum
import uuid
from datetime import datetime


class OrderStatus(str, enum.Enum):
    current = "current"
    scheduled = "scheduled"
    completed = "completed"
    incomplete = "incomplete"
    history = "history"


class ActivityStatus(str, enum.Enum):
    driver_not_assigned = "driver_not_assigned"
    pickup_initiated = "pickup_initiated"
    picked_up = "picked_up"
    delivery_initiated = "delivery_initiated"
    delivery_in_progress = "delivery_in_progress"
    delivered = "delivered"


class Order(Base):
    __tablename__ = "orders"
    __table_args__ = (
        # A date-only stop always stores 00:00, so it sorts with its day and never carries a stray time.
        CheckConstraint(
            "pickup_time_specified OR pickup_planned_at::time = '00:00'",
            name="ck_orders_pickup_date_only_midnight",
        ),
        CheckConstraint(
            "delivery_time_specified OR delivery_planned_at::time = '00:00'",
            name="ck_orders_delivery_date_only_midnight",
        ),
    )

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    order_number = Column(String, unique=True, nullable=False)

    # Driver assignment
    driver_id = Column(UUID(as_uuid=True), ForeignKey("tenants.id", ondelete="SET NULL"), nullable=True)
    vendor_id = Column(UUID(as_uuid=True), ForeignKey("tenants.id", ondelete="SET NULL"), nullable=True)

    pickup_name = Column(String)
    pickup_phone = Column(String)
    pickup_email=Column(String)
    pickup_address = Column(String)
    # Planned wall-clock time; *_time_specified False means date-only.
    pickup_planned_at = Column(DateTime(timezone=False), nullable=False)
    pickup_time_specified = Column(Boolean, nullable=False, default=True, server_default=text("true"))

    delivery_name = Column(String)
    delivery_phone = Column(String)
    delivery_email = Column(String)
    delivery_address = Column(String)
    delivery_planned_at = Column(DateTime(timezone=False), nullable=False)
    delivery_time_specified = Column(Boolean, nullable=False, default=True, server_default=text("true"))

    delivery_category_id = Column(
        UUID(as_uuid=True), ForeignKey("delivery_categories.id", ondelete="SET NULL"), nullable=True
    )
    pickup_place_id = Column(String, nullable=True)
    pickup_latitude = Column(Float, nullable=True)
    pickup_longitude = Column(Float, nullable=True)
    pickup_city_id = Column(UUID(as_uuid=True), ForeignKey("cities.id", ondelete="SET NULL"), nullable=True)
    pickup_zone_id = Column(
        UUID(as_uuid=True), ForeignKey("operational_zones.id", ondelete="SET NULL"), nullable=True
    )
    delivery_place_id = Column(String, nullable=True)
    delivery_latitude = Column(Float, nullable=True)
    delivery_longitude = Column(Float, nullable=True)
    delivery_city_id = Column(UUID(as_uuid=True), ForeignKey("cities.id", ondelete="SET NULL"), nullable=True)
    delivery_zone_id = Column(
        UUID(as_uuid=True), ForeignKey("operational_zones.id", ondelete="SET NULL"), nullable=True
    )
    route_distance_meters = Column(Integer, nullable=True)
    route_duration_seconds = Column(Integer, nullable=True)
    surcharge_ids = Column(JSON, nullable=False, default=list)
    applied_charges = Column(JSON, nullable=False, default=list)

    items = Column(JSON)

    subtotal = Column(Float, default=0)
    gst_rate = Column(Float, default=0)
    gst_amount = Column(Float, default=0)
    pst_rate = Column(Float, default=0)
    pst_amount = Column(Float, default=0)
    delivery_fees = Column(Float, default=0)
    delivery_tips = Column(Float, default=0)
    # Server-priced discount lines; `discount` is their sum.
    discount = Column(Float, default=0)
    applied_discounts = Column(JSON, nullable=False, default=list)
    coupon_code = Column(String(40), nullable=True)
    total = Column(Float, default=0)

    # Immutable payout agreed when a driver is assigned/accepts the delivery.
    driver_payout = Column(Numeric(10, 2), nullable=True)
    driver_fee_payout = Column(Numeric(10, 2), nullable=True)
    driver_tip_payout = Column(Numeric(10, 2), nullable=True)
    driver_payment_rule = Column(String(24), nullable=True)
    driver_payment_group_id = Column(
        UUID(as_uuid=True),
        ForeignKey("driver_payment_groups.id", ondelete="SET NULL"),
        nullable=True,
    )
    driver_payment_group_name = Column(String(100), nullable=True)
    driver_payment_rule_snapshot = Column(JSON, nullable=True)
    driver_payout_locked_at = Column(DateTime(timezone=True), nullable=True)

    instructions = Column(String)
    payment_method = Column(String)
    payment_details = Column(JSON)
    proof_of_delivery = Column(JSON, nullable=True)
    pickup_verification = Column(JSON, nullable=True)
    incident_report = Column(JSON, nullable=True)

    status = Column(Enum(OrderStatus, name="orderstatus_enum"), default=OrderStatus.current)
    activity_status = Column(
        Enum(ActivityStatus, name="activitystatus_enum"),
        default=ActivityStatus.driver_not_assigned,
    )
    ready_for_pickup = Column(Boolean, default=False)

    # Activity status checkpoint timestamps
    pickup_initiated_at = Column(DateTime(timezone=True), nullable=True)
    picked_up_at = Column(DateTime(timezone=True), nullable=True)
    delivery_initiated_at = Column(DateTime(timezone=True), nullable=True)
    delivery_in_progress_at = Column(DateTime(timezone=True), nullable=True)
    delivered_at = Column(DateTime(timezone=True), nullable=True)

    # Broadcast / publish state
    published = Column(Boolean, default=False, nullable=False)
    published_at = Column(DateTime(timezone=True), nullable=True)

    order_placed_time = Column(String)

    created_at = Column(DateTime(timezone=True), server_default=func.now())

    # Relationships
    driver = relationship("Tenant", foreign_keys=[driver_id], backref="orders")
    vendor = relationship("Tenant", foreign_keys=[vendor_id], backref="vendor_orders")

File: libs/shared/contracts/src/order.ts
/**
 * Order wire contract for `/api/v1/orders`. Mirrors `OrderResponse` in
 * `apps/api/app/schemas/order.py` — change both together.
 *
 * Do NOT use `OrderStatus` from `@dispatch/shared/domain`; that enum predates
 * the current API and lists a different set of values.
 */

/** Which bucket the order sits in. Derived server-side. */
export type OrderStatus = 'current' | 'scheduled' | 'completed' | 'incomplete' | 'history';

/** The driver-driven progress of a job. */
export type ActivityStatus =
  | 'driver_not_assigned'
  | 'pickup_initiated'
  | 'picked_up'
  | 'delivery_initiated'
  | 'delivery_in_progress'
  | 'delivered';

export type IncidentStage = 'pickup' | 'delivery';

export type IncidentReason =
  | 'no_answer'
  | 'wrong_address'
  | 'business_closed'
  | 'parcel_issue'
  | 'refused'
  | 'other';

/** Valid reasons per stage; the API rejects mismatches with a 400. */
export const PICKUP_INCIDENT_REASONS: readonly IncidentReason[] = [
  'no_answer',
  'wrong_address',
  'business_closed',
  'parcel_issue',
  'other',
];

export const DELIVERY_INCIDENT_REASONS: readonly IncidentReason[] = [
  'no_answer',
  'wrong_address',
  'refused',
  'other',
];

/** Reasons the API requires a non-empty description for, at either stage. */
export const INCIDENT_REASONS_REQUIRING_DESCRIPTION: readonly IncidentReason[] = [
  'other',
  'parcel_issue',
];

export type PaymentMethod = 'cash_on_delivery' | 'credit_card';

export interface OrderItem {
  itemName: string;
  itemPrice: number;
  itemQty: number;
}

export interface AppliedCharge {
  id: string | null;
  kind: 'after_hours' | 'surcharge' | 'special_occasion';
  label: string;
  amount: number;
}

export type ManualDiscountKind = 'percentage' | 'fixed_amount';

export type ManualDiscountReason =
  | 'late_delivery'
  | 'damaged_item'
  | 'wrong_address_our_fault'
  | 'sales_goodwill'
  | 'price_correction'
  | 'other';

/** A discount a dispatcher applies by hand. The server prices and caps it. */
export interface ManualDiscount {
  kind: ManualDiscountKind;
  value: number;
  reason: ManualDiscountReason;
  note?: string | null;
}

/** One priced discount line. Discounts only ever come off the delivery fee. */
export interface AppliedDiscount {
  source: 'manual' | 'automatic' | 'code';
  kind: string;
  label: string;
  value: number;
  amount: number;
  reason: string | null;
  note: string | null;
  applied_by: string | null;
}

export interface DriverInfo {
  id: string;
  name: string;
  contact_name: string | null;
  contact_phone_number: string | null;
  contact_phone_country_code: string | null;
}

export interface ProofOfDelivery {
  signature?: boolean;
  picture?: boolean;
  submission?: {
    recipient_name?: string;
    signature_path?: string;
    signature_uploaded_at?: string;
    photo_path?: string;
    photo_uploaded_at?: string;
    note?: string | null;
  };
}

export interface PickupVerification {
  method: 'qr' | 'photo';
  verified_at: string;
  verified_by: string | null;
  photo_path?: string | null;
  note?: string | null;
}

export interface IncidentReport {
  id: string;
  stage: IncidentStage;
  reason: IncidentReason;
  description: string | null;
  reported_by: string | null;
  reported_at: string;
}

/** Full order as admins and dispatchers receive it. */
export interface OrderResponse {
  id: string;
  order_number: string;
  driver_id: string | null;
  vendor_id: string | null;
  status: OrderStatus;
  activity_status: ActivityStatus;
  ready_for_pickup: boolean;

  pickup_name: string;
  pickup_phone: string;
  pickup_email: string;
  pickup_address: string;
  /** Wall-clock time, no zone ("2026-09-20T13:40:00"); read it with order-schedule helpers. */
  pickup_planned_at: string;
  /** False means date-only; the time part is 00:00. */
  pickup_time_specified: boolean;
  pickup_place_id: string | null;
  pickup_latitude: number | null;
  pickup_longitude: number | null;
  pickup_city_id: string | null;
  pickup_zone_id: string | null;

  delivery_name: string;
  delivery_phone: string;
  delivery_email: string;
  delivery_address: string;
  delivery_planned_at: string;
  delivery_time_specified: boolean;
  delivery_place_id: string | null;
  delivery_latitude: number | null;
  delivery_longitude: number | null;
  delivery_city_id: string | null;
  delivery_zone_id: string | null;

  delivery_category_id: string | null;
  /** Quoted pickup-to-drop leg, stamped at creation. Null without a quote. */
  route_distance_meters: number | null;
  route_duration_seconds: number | null;
  surcharge_ids: string[];
  applied_charges: AppliedCharge[];

  items: OrderItem[];
  subtotal: number;
  gst_rate: number;
  gst_amount: number;
  pst_rate: number;
  pst_amount: number;
  delivery_fees: number;
  delivery_tips: number;
  /** Server-priced: the sum of `applied_discounts`, never sent by a client. */
  discount: number;
  applied_discounts: AppliedDiscount[];
  coupon_code: string | null;
  total: number;

  instructions: string | null;
  order_placed_time: string | null;
  payment_method: PaymentMethod;
  payment_details: Record<string, unknown> | null;

  proof_of_delivery: ProofOfDelivery | null;
  pickup_verification: PickupVerification | null;
  incident_report: IncidentReport | null;
  driver: DriverInfo | null;

  /** Driver earnings. Locked to a snapshot once the order is accepted. */
  driver_payout: number | null;
  driver_fee_payout: number | null;
  driver_tip_payout: number | null;
  driver_payment_rule: string | null;
  driver_payment_group_id: string | null;
  driver_payment_group_name: string | null;
  driver_payout_locked_at: string | null;

  published: boolean;
  published_at: string | null;
  created_at: string | null;
}

/** Item as drivers see it; the API zeroes prices, so they're omitted here. */
export type DriverOrderItem = Omit<OrderItem, 'itemPrice'>;

/**
 * Driver-scoped order (`_driver_order_response` in `orders.py`). Platform
 * finances come back zeroed for drivers, so they're omitted rather than typed
 * as always-zero — no screen can present them as real money.
 */
export interface DriverOrder
  extends Pick<
    OrderResponse,
    | 'id'
    | 'status'
    | 'activity_status'
    | 'ready_for_pickup'
    | 'pickup_name'
    | 'pickup_phone'
    | 'pickup_address'
    | 'pickup_planned_at'
    | 'pickup_time_specified'
    | 'pickup_latitude'
    | 'pickup_longitude'
    | 'delivery_name'
    | 'delivery_phone'
    | 'delivery_address'
    | 'delivery_planned_at'
    | 'delivery_time_specified'
    | 'delivery_latitude'
    | 'delivery_longitude'
    | 'route_distance_meters'
    | 'route_duration_seconds'
    | 'instructions'
    | 'driver_payout'
    | 'driver_fee_payout'
    | 'driver_tip_payout'
    | 'driver_payment_rule'
    | 'proof_of_delivery'
    | 'pickup_verification'
    | 'incident_report'
    | 'driver'
    | 'published'
    | 'published_at'
    | 'created_at'
  > {
  order_number: string | null;
  items: DriverOrderItem[];
}
